### PR TITLE
Friends

### PR DESCRIPTION
**Pull request by AnilRedshift**
This adds the friends join table, the routes to add, remove, and view friends.

It also fixes authorization to handle this new wrinkle.

### CODE REVIEW

File: app/controllers/friends_controller.rb [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+      current_user.friends.delete(friend)
+      render json: { status: "destroyed" }
+    else
+      render json: { error: "Friend not found" }, status: 404
+    end
+  end
+
+  private
+  def friends_params
+    # TODO how to throw a 400 if the id is missing?
```
Comment (yhordi) [inline app/controllers/friends_controller.rb]:
```
params.permit(:id)
```

https://github.com/rails/strong_parameters

File: app/models/user.rb [hunk 2]
```diff
@@ -1,6 +1,16 @@
 class User < ApplicationRecord
   has_secure_password
+  has_and_belongs_to_many(:friends, class_name: :User, join_table: 'friends', foreign_key: 'user_a_id', association_foreign_key: 'user_b_id')
```
Comment (yhordi) [inline app/models/user.rb]: fix this <3

File: spec/controllers/friends_controller_spec.rb [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+describe FriendsController do
+  let(:user) { FactoryGirl.create :user }
+  let(:friend) { FactoryGirl.create :user }
+  before(:each) { user.friends << friend }
+
+  context 'when logged in as the user' do
+    before(:each) { session[:user_id] = user.id.to_s }
+
+    describe 'FoodsController#index' do
```
Comment (yhordi) [inline spec/controllers/friends_controller_spec.rb]: Stupid Atom. FriendsController :|

File: spec/models/user_spec.rb [hunk 4]
```diff
@@ -13,5 +13,31 @@
       user.books << book
       expect(user.books).to include(book)
     end
+
+    it 'can be associated with another friend' do
```
Comment (yhordi) [inline spec/models/user_spec.rb]: You can clean this up by placing it in your context block below.

File: spec/controllers/friends_controller_spec.rb (hunk 3 above)
Comment (AnilRedshift) [inline spec/controllers/friends_controller_spec.rb]: fixed

File: app/controllers/friends_controller.rb (hunk 1 above)
Comment (AnilRedshift) [inline app/controllers/friends_controller.rb]: done

File: spec/models/user_spec.rb (hunk 4 above)
Comment (AnilRedshift) [inline spec/models/user_spec.rb]: fixed

**Pull request merged**